api/holders.go

```diff
@@ -1,80 +1,93 @@
+	ctx2, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
 
-// getHolders handler function receives the token contract address from the
-// request and tries to get its token holders. If something fails, the given
-// token not exists or has not registered token holders, it will returns and api
-// error.
-func (h *holdersHandlers) getHolders(msg *api.APIdata, ctx *httprouter.HTTPContext) error {
+	// get token holders from the database
 	addr := common.HexToAddress(ctx.URLParam("address"))
-	th, err := h.scanner.GetTokenHolders(addr)
+	dbHolders, err := capi.sqlc.TokenHoldersByTokenID(ctx2,
+		queries.TokenHoldersByTokenIDParams{
+			TokenID: addr.Bytes(),
+			Limit:   -1,
+			Offset:  0,
+		})
 	if err != nil {
-		log.Errorf("error getting token with address %s: %s", addr, err)
+		// if database does not contain any token holder for this token, return
+		// not found, else return generic error.
+		if errors.Is(sql.ErrNoRows, err) {
+			log.Errorf("no holders found for address %s: %w", addr, err)
+			return ErrNoFoundTokenHolders.Withf("no holders found for address %s", addr)
+		}
+		log.Errorf("error getting token with address %s: %w", addr, err)
 		return ErrCantGetTokenHolders.Withf("error getting token with address %s", addr)
 	}
-	if th == nil {
-		log.Errorf("token with address %s", addr)
-		return ErrUnknownToken.Withf("token with address %s", addr)
-	}
-
+	// encode the response with the token holders addresses
 	holders := TokenHoldersResponse{Holders: []string{}}
-	for _, addr := range th.Holders() {
-		holders.Holders = append(holders.Holders, addr.Hex())
-	}
-	if len(holders.Holders) == 0 {
-		log.Errorf("no holders found for %s", addr)
-		return ErrNoFoundTokenHolders.Withf("no holders found for %s", addr)
+	for _, holder := range dbHolders {
+		holders.Holders = append(holders.Holders, common.BytesToAddress(holder).Hex())
 	}
 	response, err := json.Marshal(holders)
 	if err != nil {
-		log.Errorf("error marshalling holder of %s: %s", addr, err)
+		log.Errorf("error marshalling holder of %s: %w", addr, err)
 		return ErrEncodeTokenHolders.Withf("error marshalling holder of %s", addr)
 	}
 	return ctx.Send(response, api.HTTPstatusOK)
 }
+
+// countHolders handler function returns the current number of token holders
+// for the provided token address in the current database.
+// TODO: Only for debug, consider to delete
+func (capi *census3API) countHolders(msg *api.APIdata, ctx *httprouter.HTTPContext) error {
+	ctx2, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	addr := common.HexToAddress(ctx.URLParam("address"))
+	numberOfHolders, err := capi.sqlc.CountTokenHoldersByTokenID(ctx2, addr.Bytes())
+	if err != nil {
+		if errors.Is(sql.ErrNoRows, err) {
+			log.Errorf("no holders found for address %s: %w", addr, err)
+			return ErrNoFoundTokenHolders.Withf("token address: %s", addr)
```

Comment:
Do you want to treat this as an error? I think it is a bit misleading, the call in this case resolved fine but no holders were found. We can return a 200 with an empty list and a message saying "no holders found"

Reply:
Yes, I am agree. Anyway I am not sure if the current token holders endpoints are necessary or not (probably not), but I will update them before merge.

Reply:
I solve it returning `204 No content`, what do you think @jordipainan?